sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala

```diff
@@ -0,0 +1,367 @@
+   * @param outputBuffer buffer where uncompressed data will be written
+   * @param outputOffset offset in the uncompressed buffer to start writing data
+   * @param outputLength expected length of the uncompressed data in bytes
+   * @param inputBuffer buffer containing the compressed data
+   * @param inputOffset offset in the compressed buffer where compressed data starts
+   * @param inputLength length of the compressed data in bytes
+   */
+  def decompressBuffer(
+      outputBuffer: DeviceMemoryBuffer,
+      outputOffset: Long,
+      outputLength: Long,
+      inputBuffer: DeviceMemoryBuffer,
+      inputOffset: Long,
+      inputLength: Long): Unit
+
+  /**
+   * Create a batched compressor instance
+   * @param maxBatchMemorySize The upper limit in bytes of temporary and output memory usage at
+   *                           which a batch should be compressed. A single table that requires
+   *                           temporary and output memory above this limit is allowed but will
+   *                           be compressed individually.
+   * @return batched compressor instance
+   */
+  def createBatchCompressor(maxBatchMemorySize: Long): BatchedTableCompressor
+
+  /**
+   * Create a batched compressor instance
+   * @param maxBatchMemorySize The upper limit in bytes of temporary and output memory usage at
+   *                           which a batch should be decompressed. A single buffer that requires
+   *                           temporary and output memory above this limit is allowed but will
+   *                           be decompressed individually.
+   * @return batched decompressor instance
+   */
+  def createBatchDecompressor(maxBatchMemorySize: Long): BatchedBufferDecompressor
+}
+
+object TableCompressionCodec {
+  private val codecNameToId = Map(
+    "copy" -> CodecType.COPY)
+
+  /** Get a compression codec by short name or fully qualified class name */
+  def getCodec(name: String): TableCompressionCodec = {
+    val codecId = codecNameToId.getOrElse(name.toLowerCase(Locale.ROOT),
+      throw new IllegalArgumentException(s"Unknown table codec: $name"))
+    getCodec(codecId)
+  }
+
+  /** Get a compression codec by ID, using a cache. */
+  def getCodec(codecType: Byte): TableCompressionCodec = {
+    codecType match {
+      case CodecType.COPY => new CopyCompressionCodec
+      case _ => throw new IllegalArgumentException(s"Unknown codec ID")
+    }
+  }
+}
+
+/**
+ * Base class for batched compressors
+ * @param maxBatchMemorySize The upper limit in bytes of temporary and output memory usage at
+ *                           which a batch should be compressed. A single table that requires
+ *                           temporary and output memory above this limit is allowed but will
+ *                           be compressed individually.
+ */
+abstract class BatchedTableCompressor(maxBatchMemorySize: Long) extends AutoCloseable with Arm {
+  private[this] val tables = new ArrayBuffer[ContiguousTable]
```

Comment:
could we have comments for these? I am not entirely sure what `oversizedOutBuffers` is. I believe it is larger than the regular device buffer but a comment here would help.